core/shared/src/main/scala/scalaz/zio/Fiber.scala

```diff
@@ -24,11 +24,21 @@ package scalaz.zio
 trait Fiber[+E, +A] { self =>
 
   /**
-   * Joins the fiber, with suspends the joining fiber until the result of the
+   * Observes the fiber, which suspends the joining fiber until the result of the
+   * fiber has been determined.
+   */
+  def observe: IO[Nothing, ExitResult[E, A]]
+
+  /**
+   * Joins the fiber, which suspends the joining fiber until the result of the
    * fiber has been determined. Attempting to join a fiber that has been or is
    * killed before producing its result will result in a catchable error.
    */
-  def join: IO[E, A]
+  def join: IO[E, A] = observe.flatMap {
+    case ExitResult.Completed(a)   => IO.now(a)
+    case ExitResult.Failed(e, _)   => IO.fail(e)
```

Comment:
Again we're throwing away `Throwables`. I'm pretty convinced we have to change `Fail` to accept defects, too, so we can eliminate the increasing number of cases where defects are lost. Of course, by default, the list of throwables can be `Nil` so the interface can remain the same.

Reply:
Yes, that's annoying. I have added `IO.fail0` for internal use only, let me know what you think.